internal/painter/software/painter_test.go

```diff
@@ -12,37 +12,65 @@ import (
 	"fyne.io/fyne/theme"
 )
 
-func makeTestImage() image.Image {
-	src := image.NewNRGBA(image.Rect(0, 0, 3, 3))
-
-	for y := 0; y < 3; y++ {
-		for x := 0; x < 3; x++ {
-			if x == y {
-				src.Set(x, y, color.Black)
-			} else {
-				src.Set(x, y, color.White)
-			}
+func makeTestImage(w, h int) image.Image {
+	src := image.NewNRGBA(image.Rect(0, 0, w, h))
+	colors := []color.Color{color.White, color.Black}
+	c := 0
+	for y := 0; y < h; y++ {
+		c = y % 2
+		for x := 0; x < w; x++ {
+			src.Set(x, y, colors[c])
+			c = 1 - c
 		}
 	}
-
 	return src
 }
 
 func TestPainter_PaintImage(t *testing.T) {
+	img := canvas.NewImageFromImage(makeTestImage(3, 3))
+
 	c := test.NewCanvas()
 	c.SetPadded(false)
-	c.SetContent(canvas.NewImageFromImage(makeTestImage()))
+	c.SetContent(img)
 	c.Resize(fyne.NewSize(50, 50))
 	p := software.NewPainter()
 
 	target := p.Paint(c)
 	test.AssertImageMatches(t, "draw_image_default.png", target)
 }
 
+func TestPainter_PaintImageScalePixels(t *testing.T) {
+	img := canvas.NewImageFromImage(makeTestImage(3, 3))
+	img.ScaleMode = canvas.ImageScalePixels
+
+	c := test.NewCanvas()
+	c.SetPadded(false)
+	c.SetContent(img)
+	c.Resize(fyne.NewSize(50, 50))
+	p := software.NewPainter()
+
+	target := p.Paint(c)
+	test.AssertImageMatches(t, "draw_image_ImageScalePixels.png", target)
+}
+
+func TestPainter_PaintImageScaleSmooth(t *testing.T) {
+	img := canvas.NewImageFromImage(makeTestImage(3, 3))
+	img.ScaleMode = canvas.ImageScaleSmooth
+
+	c := test.NewCanvas()
+	c.SetPadded(false)
+	c.SetContent(img)
+	c.Resize(fyne.NewSize(50, 50))
+	p := software.NewPainter()
+
+	target := p.Paint(c)
+	test.AssertImageMatches(t, "draw_image_ImageScaleSmooth.png", target)
+}
+
 func TestPainter_PaintImage_StretchX(t *testing.T) {
 	c := test.NewCanvas()
 	c.SetPadded(false)
-	c.SetContent(canvas.NewImageFromImage(makeTestImage()))
+	c.SetContent(canvas.NewImageFromImage(makeTestImage(4, 3)))
```

Comment:
Just scanning through the images and this new 4-3 ratio is hard to tell if it is stretched in the 1-2 ratio of the output image (same below). Can you go back to the 3-3 please?

Reply:
Sure